Make logged in message permanent in users page

sharidas opened the pull request:
Currently we are storing the data as
session. This would help to know if the
impersonation happened in the users page.
Once the user is logged out the localstorage
data is removed.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
The reason for adding this change is, currently we store the olduserid in the session. And hence from the JS side in the users page the message: "logged in as xyz" was not shown. In the core it was being hidden by the users.js and groups.js. So I found this as one of the solution to set the data in the localstorage and read  and then remove it when user logs out back to the orignal user.

PVince81 commented on js/impersonate_logout.js:
```diff
@@ -26,6 +22,7 @@ $(document).ready(function () {
 		).promise();
 
 		promisObj.done(function () {
+			localStorage.removeItem('oldUserId');
```
if the session times out this will never be called

PVince81 commented:
I'm not too happy to use local storage here. If the session times out the local storage value will still be there as it's not tied to the session.

Now I'm not sure how difficult it would be to expose the session value to the client side through html for the JS code to be able to read this. So far I'm not aware of any hooks that could allow this.

There is one location in the code that usually exposes session stuff to the client and that is this file: https://github.com/owncloud/core/blob/anon-upload/core/js/config.php#L206

Apparently there is already a hook there that could allow apps to add values there on PHP level.
I think you can register to this hook in the impersonate app's "app.php".

sharidas commented:
@PVince81  I have verified this change with the change I have made in the core for the users.js, groups.js and deleteHandler.js

PVince81 commented on controller/logoutcontroller.php:
```diff
@@ -58,4 +59,13 @@ public function logoutcontroller($userid) {
+	 * @param array $data
+	 */
+	public function getJavaScript(array $data) {
+		$data['array']['oc_oldUserId'] = json_encode([
```
only add if not null ?

sharidas replied:
Updated in my latest change.

sharidas commented:
The only problem I see now is when the user navigates to the users page, the "Logged in as " message moves a bit left to allow notification message to show up. And once the notification disappears, it will come back to the proper position. Now its all with some css work to arrange the bits. Need to figure out how to handle this. 
The reason why I did this change is because of the hides/unhides happening in users.js and groups.js were causing problem with the notification we want to push for impersonated user.

sharidas commented:
I have updated the latest change. Fixed the css issue. In the latest change I have removed OC_Hook. The reason for that is, we do get the data from impersonate-logout.js. The css issues which caused the problem
a) When user lands up in the users page, Logged in message comes up and then goes off. Has now been fixed.
b) When user tries to create a user and then deletes the user, the 2 notification messages are now in the same row. This issue is also fixed. 

The Logged in message is permanent until user is logged out to the orignal user.

PVince81 commented on js/impersonate.js:
```diff
@@ -71,6 +71,13 @@
+				$("#notification").css("margin-left", "2px");
+				$("#impersonate-notification").css("margin-right", "2px");
+				$("#impersonate-notification").css("vertical-align", "top");
+				var $deleteHandler = oldInitDeleteHandling.apply(this, arguments);
```
did you maybe forget to return something ?

PVince81 commented on js/impersonate_logout.js:
```diff
@@ -2,22 +2,19 @@ $(document).ready(function () {
-	var text = t(
-		'core',
-		'<a href="{docUrl}">{displayText}</a>',
+	var text = t( 'core',
```
don't translate HTML

PVince81 commented on js/impersonate_logout.js:
```diff
@@ -2,22 +2,19 @@ $(document).ready(function () {
+		'</div>',
 		{
 			docUrl: OC.generateUrl('apps/files'),
 			displayText: "Logged in as " + OC.getCurrentUser().uid,
```
this is what needs translating with parameters

sharidas commented:
Updated the return for initDeleteHandling. 
Translated only the displayText. The html remains untranslated.

PVince81 commented on js/impersonate.js:
```diff
@@ -71,6 +71,14 @@
+			var oldInitDeleteHandling = obj.initDeleteHandling;
+			obj.initDeleteHandling = function () {
+				$("#notification").css("margin-left", "2px");
+				$("#impersonate-notification").css("margin-right", "2px");
```
can we use CSS classes instead of hard-coding styles in code ?
the same applies for the notification element above.

Please define CSS classes in the impersonate app and set/unset them to the elements

PVince81 commented on js/impersonate_logout.js:
```diff
@@ -2,22 +2,15 @@ $(document).ready(function () {
+	var text = '<div id="impersonate-notification" ' +
+		'style="display: inline-block;background-color: #fc4;margin: 0 auto;max-width: 60%; z-index: 8000;border: 0;padding: 1px 8px;position: relative;top: 0;border-bottom-left-radius: 3px;border-bottom-right-radius: 3px; -ms-filter: \'progid:DXImageTransform.Microsoft.Alpha(Opacity=90)\';opacity: .9;">' +
+		'<div class="row">' +
+		'<a href="' + docUrl + '" style="text-align: center;">' + displayText + '</a>' +
```
If the uid contained HTML code one could inject it here. Please escape it.

Even better would be to use a Handlebars template, see https://github.com/owncloud/core/blob/master/core/js/sharedialogview.js#L16

sharidas commented:
Used Handlebar template in the code. Tried to replace the css functions by having a css file and added changes in the css file for the app.

PVince81 commented on appinfo/app.php:
```diff
@@ -10,8 +10,10 @@
+
 if(\OC::$server->getSession()->get('oldUserId') !== null) {
 	\OCP\Util::addScript('impersonate','impersonate_logout');
+	\OCP\Util::addStyle('impersonate', 'impersonate_logout');
```
I think a generic "impersonate.css" would be fine too

PVince81 commented on css/impersonate_logout.css:
```diff
@@ -0,0 +1,21 @@
+	top: 0;
+	border-bottom-left-radius: 3px;
+	border-bottom-right-radius: 3px;
+	-ms-filter: 'progid:DXImageTransform.Microsoft.Alpha(Opacity=90)';
```
is that still needed for IE11 ?

sharidas replied:
face palm :( I should have looked into this. Reference: https://msdn.microsoft.com/en-us/library/ms530752(v=vs.85).aspx Its no longer available in IE10. I will remove it.

PVince81 commented on js/impersonate.js:
```diff
@@ -71,6 +71,11 @@
+			var oldInitDeleteHandling = obj.initDeleteHandling;
+			obj.initDeleteHandling = function () {
+				var $deleteHandler = oldInitDeleteHandling.apply(this, arguments);
+				return $deleteHandler;
```
Overriding without doing anything is pointless.

Didn't you have code here that changed the styles ?

sharidas replied:
I moved the changes to the css. Will remove this code.

PVince81 commented on js/impersonate_logout.js:
```diff
@@ -2,22 +2,42 @@ $(document).ready(function () {
+			var baseTemplate = this._getTemplate('base', TEMPLATE_BASE);
+			return baseTemplate({
+				docUrl: OC.generateUrl('apps/files'),
+				displayText: t('core',"Logged in as " + OC.getCurrentUser().uid)
```
this won't work with the translation system.

Please use a placeholder string, see https://github.com/owncloud/core/blob/v10.0.0/core/js/setupchecks.js#L281

sharidas commented:
1) I have removed the code from impersonate.js ( no longer needed, because I have moved the css changes to impersonate.css file ).
2) Renamed the css file impersonate.css.
3) Used placeholder for translation support.
4) Removed -ms-filter from the css as its removed from IE10.

PVince81 commented:
Code looks good now :+1:

If it works correctly with users page / delete operation, etc, feel free to merge

sharidas commented:
I have verified as follows:
- [x] Tested by creating a user and deleting it in the users page
- [x] Created multiple users and deleting one by one quickly in the users page
- [x] Clicked the "Everyone" link multiple times and repeated steps mentioned in (1) and (2) checkbox
- [x] Clicked the groups link in the left panel in the users page and repeated steps mentioned in (1) and (2) checkbox.
- [x] Verified behaviour as admin and subadmin user respectively.
